### PR TITLE
Limit compilation time.

### PR DESCRIPTION
**Pull request by twd2**

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
+sudo: required
+dist: xenial
+env:
+  - CXX=g++-4.8 TRAVIS_NODE_VERSION="6.9.4"
```
Comment (iceboy233) [inline .travis.yml]: ???

File: .travis.yml [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+language: python
+python:
+  - "3.5"
+  - "3.5-dev"
+  - "nightly"
+install:
+  - apt install libseccomp-dev linux-libc-dev
```
Comment (iceboy233) [inline .travis.yml]: ???

File: jd4/case.py [hunk 3]
```diff
@@ -126,7 +88,7 @@ def do_stdout(self, stdout_file):
 
 class APlusBCase(CaseBase):
     def __init__(self, a, b):
-        super().__init__(1000000000, DEFAULT_MEM_KB * 1024, PROCESS_LIMIT, 10)
+        super().__init__(DEFAULT_TIME_MS * 1e6, DEFAULT_MEM_KB * 1024, PROCESS_LIMIT, 10)
```
Comment (iceboy233) [inline jd4/case.py]: 1e6 is float. You should write 1000000.

File: jd4/daemon.py [hunk 4]
```diff
@@ -39,20 +42,23 @@ def __init__(self, session, request, ws):
                 else:
                     raise SystemError('Unknown event: {}'.format(event))
             else:
-                type = self.request.pop('type')
-                if type == 0:
+                self.type = self.request.pop('type')
```
Comment (iceboy233) [inline jd4/daemon.py]: move this into `__init__`.

also note that the `problem_data_change` event should not be handled by `JudgeHandler`.

File: jd4/daemon.py [hunk 5]
```diff
@@ -103,12 +110,16 @@ def __init__(self, session, request, ws):
         total_memory_usage_bytes = 0
         for index, case in enumerate(cases):
             status, score, time_usage_ns, memory_usage_bytes, stderr = await case.judge(sandbox, package)
+            if self.type == 1:
```
Comment (iceboy233) [inline jd4/daemon.py]: could you write like this:

```
case = {...}
if self.type == 1:
    case['judge_text'] = ...
```

File: jd4/cgroup.py [hunk 6]
```diff
@@ -15,6 +15,11 @@
 MEMORY_CGROUP_ROOT = '/sys/fs/cgroup/memory/jd4'
 PIDS_CGROUP_ROOT = '/sys/fs/cgroup/pids/jd4'
 
+WAIT_JITTER_NS = 5000000
+PROCESS_LIMIT = 32
```
Comment (iceboy233) [inline jd4/cgroup.py]: These default limits should not be put here and imported.

File: jd4/langs.py [hunk 7]
```diff
@@ -40,7 +53,11 @@ def _load_langs():
                                 lang_config['code_file'],
                                 lang_config['execute_file'],
                                 shlex.split(lang_config['execute_args']))
-            langs[lang_name] = partial(_compiler_build, compiler)
+            langs[lang_name] = partial(
+                _compiler_build, compiler,
+                time_limit_ns=lang_config.get('time_limit_ms', DEFAULT_TIME_MS) * 1e6,
```
Comment (iceboy233) [inline jd4/langs.py]: 1e6 is float.

File: jd4/case.py [hunk 8]
```diff
@@ -1,65 +1,27 @@
...
-from os import cpu_count, mkfifo, path
+from os import mkfifo, path
 from random import randint
 from shutil import copyfileobj
 from zipfile import ZipFile
 
 from jd4._compare import compare_stream
-from jd4.cgroup import CGroup, try_init_cgroup
+from jd4.cgroup import CGroup, try_init_cgroup, accept_and_limit, \
+                       PROCESS_LIMIT, DEFAULT_TIME_MS, DEFAULT_MEM_KB
```
Comment (iceboy233) [inline jd4/case.py]: These constants should be put in file rather than imported.

File: examples/langs.yaml [hunk 9]
```diff
@@ -5,27 +5,31 @@ c:
     code_file: foo.c
     execute_file: foo
     execute_args: foo
+    time_limit_ms: 20000
```
Comment (iceboy233) [inline examples/langs.yaml]: Let's use default value and not adding it here.

**Review by iceboy233 — CHANGES_REQUESTED** [review]

**Comment by twd2** [comment]
@iceb0y

File: jd4/daemon.py [hunk 10]
```diff
@@ -24,30 +27,33 @@ def __init__(self, session, request, ws):
         self.request = request
         self.ws = ws
         self.tag = self.request.pop('tag', None)
+        self.event = self.request.pop('event', None)
```
Comment (iceboy233) [inline jd4/daemon.py:30]: Move this event handling out of judge handler. i.e. you can add an EventHandler, though the naming is weird.

File: jd4/daemon.py [hunk 11]
```diff
@@ -103,12 +109,16 @@ def __init__(self, session, request, ws):
...
         for index, judge_task in enumerate(judge_tasks):
             status, score, time_usage_ns, memory_usage_bytes, stderr = await judge_task
+            case = {'status': status,
+                    'score': score,
+                    'time_ms': time_usage_ns // 1000000,
+                    'memory_kb': memory_usage_bytes // 1024}
+            if self.type == 1:
+                case['judge_text'] = stderr.decode(encoding='utf-8', errors='replace')
+            else:
+                case['judge_text'] = ''
```
Comment (iceboy233) [inline jd4/daemon.py:119]: I thought judge_text was optional.
Since it's mandatory, let's revert this ^_^T_T

File: jd4/pool.py [hunk 12]
```diff
@@ -54,7 +74,11 @@
                                 lang_config['code_file'],
                                 lang_config['execute_file'],
                                 shlex.split(lang_config['execute_args']))
-            _langs[lang_name] = partial(_compiler_build, compiler)
+            _langs[lang_name] = partial(
+                _compiler_build, compiler,
+                time_limit_ns=lang_config.get('time_limit_ms', DEFAULT_TIME_MS) * 1000000,
```
Comment (iceb0y) [inline jd4/pool.py]: please put these into _compiler_build()

**Pull request merged**